Bootloader/bootloader_core.c

```diff
@@ -0,0 +1,514 @@
+            {
+                // save the current page in flash memory
+                LuosBootloader_SaveLastData();
+
+                // send ack to the Host
+                LuosBootloader_SendResponse(BOOTLOADER_BIN_END_RESP);
+
+                // go to BIN_END state
+                LuosBootloader_SetState(BOOTLOADER_CRC_TEST_STATE);
+            }
+            break;
+
+        case BOOTLOADER_CRC_TEST_STATE:
+
+            if (bootloader_cmd == BOOTLOADER_CRC_TEST)
+            {
+                uint8_t crc = compute_crc();
+                // send ack to the Host
+                LuosBootloader_SendCrc(BOOTLOADER_CRC_RESP, crc, sizeof(uint16_t));
+
+                // go to READY state
+                LuosBootloader_SetState(BOOTLOADER_READY_STATE);
+            }
+            break;
+
+        case BOOTLOADER_STOP_STATE:
+            // save boot_mode in flash
+            LuosHAL_SetMode(APPLICATION_MODE);
+            // wait for the command to be send to all nodes
+            tickstart = LuosHAL_GetSystick();
+            while ((LuosHAL_GetSystick() - tickstart) < 1000)
+                ;
+            // reboot the node
+            LuosHAL_Reboot();
+            break;
+
+        default:
+            break;
+    }
+}
+#endif
+
+#ifdef BOOTLOADER_CONFIG
+/******************************************************************************
+ * @brief bootloader app
+ * @param None
+ * @return None
+ ******************************************************************************/
+void LuosBootloader_Run(void)
+{
+    switch (LuosBootloader_GetMode())
+    {
+        case APPLICATION_MODE:
+            // boot the application programmed in dedicated flash partition
+            LuosBootloader_DeInit();
+            LuosBootloader_JumpToApp();
+            break;
+
+        case BOOTLOADER_MODE:
+        default:
+            LuosBootloader_Task();
+            break;
+    }
+}
+#endif
+
+/******************************************************************************
+ * @brief Message handler called from luos library
+ * @param data received from luos network
+ * @return None
+ ******************************************************************************/
+void LuosBootloader_MsgHandler(msg_t *input)
+{
+    bootloader_cmd = input->data[0];
+
+    switch (bootloader_cmd)
+    {
+#ifndef BOOTLOADER_CONFIG
+        case BOOTLOADER_START:
+            // We're in the app,
+            // set bootloader mode, save node ID and reboot
+            LuosHAL_SetMode(BOOTLOADER_MODE);
+            LuosBootloader_SaveNodeID();
+            LuosHAL_Reboot();
+            break;
+#else
+        case BOOTLOADER_STOP:
+        case BOOTLOADER_READY:
+        case BOOTLOADER_BIN_CHUNK:
+        case BOOTLOADER_BIN_END:
+        case BOOTLOADER_CRC_TEST:
+            // we're in the bootloader,
+            // process cmd and data
+            bootloader_data_size = input->header.size - sizeof(char);
```

Comment:
Why don't you call LuosBootloader_Task here instead of using global vars to deal with it?
For now, If you receive 2 messages before LuosBootloader_Task, you are dead.
That is probably why you have to send ack messages for each LuosBootloader_Task execution...

Reply:
The acknowledge system is not an effect caused by a problem of reception, this is a system design choice. But you're right, if the bootloader receives 2 messages before the LuosBootloader_Task catch the first, we just not process it. The design of the CLI prevents this problem.
I think it's possible to find an another way to handle this but this is the simpliest and more convenient solution I found. Maybe  we could improve this in a future release ?

Reply:
I re-wrote the code to process messages in message handler, it's not possible to miss a message anymore.